caddytls/ctmonitor.go

```diff
@@ -0,0 +1,298 @@
+func getCaddyCerts() ([]string, map[string]struct{}) {
+	var (
+		caddyCerts = make(map[string]struct{}) //caddyCerts consists of the certificates that Caddy is serving.
+		caddyDNS   = make([]string, 0, 10)     //caddyDNS consists of the Subject Alternate Names that Caddy is hosting.
+	)
+	for _, inst := range caddy.Instances() {
+		inst.StorageMu.RLock()
+		certCache, ok := inst.Storage[CertCacheInstStorageKey].(*certificateCache)
+		inst.StorageMu.RUnlock()
+		if !ok || certCache == nil {
+			continue
+		}
+		certCache.RLock()
+		for _, certificate := range certCache.cache {
+			caddyDNS = append(caddyDNS, certificate.Names...)
+			certBytes := string(certificate.Certificate.Certificate[0])
+			if _, ok := caddyCerts[certBytes]; !ok {
+				caddyCerts[certBytes] = struct{}{}
+			}
+		}
+		certCache.RUnlock()
+	}
+	return caddyDNS, caddyCerts
+}
+
+func getLatestIndex(fileName string) (int, error) {
+	fmt.Printf("ct_id FilePath: %v\n", fileName)
+	indexBytes, err := ioutil.ReadFile(fileName)
+	if os.IsNotExist(err) {
+		log.Println("getLatestIndex failed, could not find the file")
+		return 0, nil
+	}
+	if err != nil {
+		return 0, err
+	}
+	indexStr := strings.TrimSpace(string(indexBytes))
+	index, err := strconv.Atoi(indexStr)
+	if err != nil {
+		log.Println("Error converting file string to int.")
+		return 0, err
+	}
+	return index, nil
+}
+
+func loadConfig() (config CtConfig) {
+	configJSON, err := os.Open(ctConfigFile)
+	if err != nil {
+		log.Printf("loadConfig error: %v", err)
+	}
+	defer configJSON.Close()
+	//var config Config
+	err = json.NewDecoder(configJSON).Decode(&config)
+	if err != nil {
+		log.Printf("[NOTICE] jsonDecode failed, error: %v\nUsing default values", err)
+		config.IncludeSubdomains = false
+		config.IncludeWildCards = false
+	}
+	return
+}
+
+// lookUpNames queries the certSpotter service for each SAN that Caddy is hosting
+// It then adds them to a set and returns a map of each certificate  mapped to a string
+// that contains identifying information for the cert.
+func lookUpNames(caddyCertsSANs []string, config QueryConfig) (map[string]string, int) {
+	// retrievedCerts is the bytes of a certificate mapped to the
+	// ID, issuer name, and before/after values.
+	retrievedCerts := make(map[string]string)
+
+	// biggestID is the most recent certificate id returned from CertSpotter.
+
+	// timeToWait is the amount of time you need to wait before querying again.
+	var biggestID, timeToWait int
+
+	// retryAfter is the timeToWait value from the response headers.
+	var retryAfter string
+
+	// If the caddyCertsSANs is empty, it shouldn't run this at all.
+	for _, domainName := range caddyCertsSANs {
+		retryAfter = queryDomainName(domainName, config, &biggestID, retrievedCerts)
+	}
+	err := putLatestID(biggestID, filePath)
+	if err != nil {
+		//return fmt.Errorf("writing latest ID: %v", err)
+		log.Printf("[WARNING] writing latest ID (%v): %v", biggestID, err)
+	}
```

Comment:
Are you sure you want the code to keep running if there is an error? Check all error handling occurrences.

Reply:
I made it so that the process will terminate after comparing all of the certificates that they found.